### DISCUSSION TITLE
Roslyn Analyzer Discussion

### DISCUSSION THREAD

**Discussion by ketanpkolte** [dotnet/BenchmarkDotNet#2666] [General]
Discussion regarding rules of diagnostics for BDN, new rules and Code samples

**Comment by ketanpkolte** [comment]
Currently I am writing a Roslyn analyzer for BDN and would like to know about some good Rules that I can apply for diagnostics.

Some Rules that I have written till now are: 
Classes containing [Benchmark] must be UnSealed and Public
Methods attributed with [Benchmark] must be parameterless

**Reply by timcassell** [reply]
> Methods attributed with [Benchmark] must be parameterless

This is not true. Benchmarks may contain parameters, but they must have the `[Arguments]` or `[ArgumentsSource]` attribute. https://benchmarkdotnet.org/articles/features/parameterization.html#sample-introarguments

**Comment by ketanpkolte** [comment]
I would like to know more about benchmarking Generic methods, since BDN allows to benchmark them only if certain criteria are met.
I wanted to confirm these two examples:  

'''public class InvalidGenericMethodBenchmark
 {
     [Params(100, 200, 300)]
     public int N;

     [Benchmark]
     public void BenchmarkInvalidGenericMethod()
     {
         InvalidGenericMethod(N);
     }

     public void InvalidGenericMethod<T>(T value) where T : new()
     {
         T instance = new T();
     }
 }'''

 '''public class ValidGenericMethodBenchmark
 {
     [Params(100, 200, 300)]
     public int N;

     [Benchmark]
     public void BenchmarkGenericMethod()
     {
         GenericMethod<int>(N);
     }

     public void GenericMethod<T>(T value)
     {
         // Simulate some work
         for (int i = 0; i < 1000; i++)
         {
             var temp = value;
         }
     }
 }'''

**Reply by timcassell** [reply]
Currently, methods annotated with `[Benchmark]` attribute may not be generic. Generic classes may have benchmarks, as long as the class is annotated with `[GenericTypeArguments]` attribute. https://benchmarkdotnet.org/articles/samples/IntroGenericTypeArguments.html

We do not care if other methods are generic that are called by the benchmark method.

**Comment by timcassell** [comment]
I recommend reading through the [documentation](https://benchmarkdotnet.org/articles/overview.html) and (especially) the existing validators source code.

**Comment by ketanpkolte** [comment]
I went through the documents mentioned i wanted to know about possible invalid genericTypeAttribute, as mentioned in my code example

**Reply by timcassell** [reply]
The number of types in the attribute must match the number of generic type arguments on the type. I think that's the only validation necessary for it.

[Edit] And a generic class containing benchmarks must have at least one of that attribute.

**Comment by silkfire** [comment]
@ketanpkolte @timcassell  Hi, I just wanted to check in and ask how things are progressing with the analyzers?

I have been working on creating a substantial set of Roslyn analyzers since April this year and have come quite far already. Everything is covered with comprehensive tests and supporting all the frameworks that BenchmarkDotNet currently supports.

Would it be okay for me to push a PR?

**Reply by timcassell** [reply]
Yes please send the PR. 🙂

**Reply by silkfire** [reply]
Done! https://github.com/dotnet/BenchmarkDotNet/pull/2837

**Comment by ketanpkolte** [comment]
@silkfire I have not made any significant progress on analyzers, since quite a while.   Regarding PR @timcassell would give better answer/guidance
